dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java

```diff
@@ -0,0 +1,331 @@
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class responsible of setting Elasticsearch mapping for content type fields
+ * @author nollymar
+ */
+public class ESMappingUtilHelper {
+
+    private ContentTypeAPI contentTypeAPI;
+    private ESMappingAPIImpl esMappingAPI;
+    private RelationshipAPI relationshipAPI;
+
+    private static class SingletonHolder {
+
+        private static final ESMappingUtilHelper INSTANCE = new ESMappingUtilHelper();
+    }
+
+    public static ESMappingUtilHelper getInstance() {
+        return ESMappingUtilHelper.SingletonHolder.INSTANCE;
+    }
+
+    private ESMappingUtilHelper() {
+        contentTypeAPI = APILocator.getContentTypeAPI(APILocator.systemUser());
+        esMappingAPI = new ESMappingAPIImpl();
+        relationshipAPI = APILocator.getRelationshipAPI();
+    }
+
+    @VisibleForTesting
+    ESMappingUtilHelper(final ContentTypeAPI contentTypeAPI, final ESMappingAPIImpl esMappingAPI,
+            final RelationshipAPI relationshipAPI) {
+        this.contentTypeAPI = contentTypeAPI;
+        this.esMappingAPI = esMappingAPI;
+        this.relationshipAPI = relationshipAPI;
+    }
+
+    /**
+     * Sets a custom index mapping for relationships and also for mapping defined on field variables
+     * using `esCustomMapping` property
+     *
+     * @param indexName - Index where mapping will be updated
+     */
+    @CloseDBIfOpened
+    public void addCustomMapping(final String indexName) {
+
+        final Set<String> mappedFields = addCustomMappingFromFieldVariables(indexName);
+
+        addCustomMappingForRelationships(indexName, mappedFields);
+
+        addMappingForRemainingFields(indexName, mappedFields);
+    }
+
+    /**
+     * Sets a mapping for all relationships except for those that contains its custom mapping using
+     * field variables
+     *
+     * @param indexName - Index where mapping will be updated
+     * @param mappedFields - Collection that contains the fields with a specific mapping until now.
+     * </br> When a put mapping request is sent to Elasticsearch for each relationship (if needed),
+     * a new entry is added to the <b>mappedFields</b> collection
+     */
+    private void addCustomMappingForRelationships(final String indexName,
+            final Set<String> mappedFields) {
+        final List<Relationship> relationships = relationshipAPI.dbAll();
+
+        for (final Relationship relationship : relationships) {
+            final String relationshipName = relationship.getRelationTypeValue().toLowerCase();
+            if (!mappedFields.contains(relationshipName)) {
+                final JSONObject properties = new JSONObject();
+                try {
+                    properties.put("properties", new JSONObject()
```

Comment:
Codacy found an issue: [The String literal "properties" appears 5 times in this file; the first occurrence is on line 97](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5739222)